File: W03_Inheritance-Polymorphism-AbstactBaseClass_Tutorial/1_Rekening/RekeningGiro.cpp
// Nama/NIM : Jonathan Yudi Gunawan/13518084
// Hari, Tanggal : Kamis, 6 Februari 2020
// Materi : Inheritance

#define fori(i,n)    for(int i = 0; i < n; i++)
#define forii(i,s,e) for(int i = s; i < e; i++)
#define max(a,b)	 a>b?a:b
#define min(a,b)	 a<b?a:b

#include "RekeningGiro.h" // include the Account header file

// Konstruktor menginisialisi saldo (parameter 1) dan suku bunga (parameter 2)
// Set suku bunga 0.0 apabila suku bunga bernilai negatif
RekeningGiro::RekeningGiro(double saldo, double sukuBunga) : Rekening(saldo), sukuBunga(max(0.0, sukuBunga)) {}

// Getter, Setter
void RekeningGiro::setSukuBunga(double sukuBunga){
	this->sukuBunga=sukuBunga;
}
double RekeningGiro::getSukuBunga() const {
	return this->sukuBunga;
}

// Member Function
// Bunga dihitung dari saldo dikali suku bunga
double RekeningGiro::hitungBunga(){
	return this->getSaldo()*this->getSukuBunga();
}

File: W07_OODesign/3_Rekening/RekeningGiro.java
class RekeningGiro extends Rekening{
	private boolean isPenatied;
	public RekeningGiro(String nama, double saldo){
		super(nama, saldo, 0.07);
		this.isPenatied = this.penalti();
	}

	public void setor(double saldo){
		this.saldo += saldo - 0.1;
	}

	public void tarik(double saldo){
		this.saldo -= saldo + 0.1;
		if(this.saldo < 0){
			this.saldo += saldo + 0.1;
		} else {
			this.isPenatied |= this.penalti();
		}
	}

	public void update(){
		if(this.isPenatied){
			this.saldo -= 10;
		}
		this.saldo += this.sukuBunga * this.saldo - this.hitungBiaya();
		this.saldo = Math.max(0, this.saldo);
	}

	public boolean penalti(){
		return this.saldo < 500;
	}
}